File: rust/llmix-rs/src/adaptive_semaphore.rs
use crate::{AdaptiveSemaphoreClosedError, LlmixError, LlmixResult};
use std::collections::{HashMap, VecDeque};
use std::sync::Mutex;
use tokio::sync::oneshot;

pub const DEFAULT_INITIAL: usize = 32;
pub const DEFAULT_MIN_CONCURRENCY: usize = 4;
pub const HEADER_BACKOFF_THRESHOLD: f64 = 0.10;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct RateLimitHeaders {
    pub remaining: usize,
    pub limit: usize,
}

#[derive(Debug)]
struct AdaptiveSemaphoreState {
    window: usize,
    available: usize,
    waiters: VecDeque<oneshot::Sender<Result<(), AdaptiveSemaphoreClosedError>>>,
    has_header_signal: bool,
    permits_to_absorb: usize,
    closed: bool,
}

#[derive(Debug)]
pub struct AdaptiveSemaphore {
    max: usize,
    min: usize,
    state: Mutex<AdaptiveSemaphoreState>,
}

#[derive(Debug)]
pub struct AdaptiveSemaphorePermit<'a> {
    semaphore: &'a AdaptiveSemaphore,
    released: bool,
}

impl AdaptiveSemaphore {
    pub fn new(initial: usize, min_concurrency: usize) -> LlmixResult<Self> {
        if initial < 1 {
            return Err(LlmixError::InvalidAdaptiveSemaphoreConfig(format!(
                "initial must be >= 1, got {initial}"
            )));
        }
        if min_concurrency < 1 {
            return Err(LlmixError::InvalidAdaptiveSemaphoreConfig(format!(
                "min_concurrency must be >= 1, got {min_concurrency}"
            )));
        }
        if initial < min_concurrency {
            return Err(LlmixError::InvalidAdaptiveSemaphoreConfig(format!(
                "initial ({initial}) must be >= min_concurrency ({min_concurrency})"
            )));
        }

        Ok(Self {
            max: initial,
            min: min_concurrency,
            state: Mutex::new(AdaptiveSemaphoreState {
                window: initial,
                available: initial,
                waiters: VecDeque::new(),
                has_header_signal: false,
                permits_to_absorb: 0,
                closed: false,
            }),
        })
    }

    pub fn with_defaults() -> Self {
        Self::new(DEFAULT_INITIAL, DEFAULT_MIN_CONCURRENCY)
            .expect("default adaptive semaphore configuration must be valid")
    }

    pub fn window(&self) -> usize {
        self.state
            .lock()
            .expect("adaptive semaphore mutex poisoned")
            .window
    }

    pub fn max_concurrency(&self) -> usize {
        self.max
    }

    pub fn min_concurrency(&self) -> usize {
        self.min
    }

    pub fn closed(&self) -> bool {
        self.state
            .lock()
            .expect("adaptive semaphore mutex poisoned")
            .closed
    }

    pub fn rebind(&self) {
        let mut state = self
            .state
            .lock()
            .expect("adaptive semaphore mutex poisoned");
        state.available = state.window;
        state.waiters.clear();
        state.permits_to_absorb = 0;
    }

    pub async fn acquire(&self) -> Result<(), AdaptiveSemaphoreClosedError> {
        loop {
            let receiver = {
                let mut state = self
                    .state
                    .lock()
                    .expect("adaptive semaphore mutex poisoned");

                if state.closed {
                    return Err(AdaptiveSemaphoreClosedError);
                }

                if state.available > 0 {
                    state.available -= 1;
                    return Ok(());
                }

                let (tx, rx) = oneshot::channel();
                state.waiters.push_back(tx);
                rx
            };

            match receiver.await {
                Ok(result) => return result,
                Err(_) => {
                    if self.closed() {
                        return Err(AdaptiveSemaphoreClosedError);
                    }
                }
            }
        }
    }

    pub async fn acquire_guard(
        &self,
    ) -> Result<AdaptiveSemaphorePermit<'_>, AdaptiveSemaphoreClosedError> {
        self.acquire().await?;
        Ok(AdaptiveSemaphorePermit {
            semaphore: self,
            released: false,
        })
    }

    pub fn close(&self) {
        let waiters = {
            let mut state = self
                .state
                .lock()
                .expect("adaptive semaphore mutex poisoned");
            state.closed = true;
            state.waiters.drain(..).collect::<Vec<_>>()
        };

        for waiter in waiters {
            let _ = waiter.send(Err(AdaptiveSemaphoreClosedError));
        }
    }

    pub fn release(&self) {
        let mut state = self
            .state
            .lock()
            .expect("adaptive semaphore mutex poisoned");

        if state.permits_to_absorb > 0 {
            state.permits_to_absorb -= 1;
            return;
        }

        if wake_waiter(&mut state.waiters) {
            return;
        }

        if state.available < state.window {
            state.available += 1;
        }
    }

    pub fn on_success(&self) {
        let mut state = self
            .state
            .lock()
            .expect("adaptive semaphore mutex poisoned");
        if state.closed || state.has_header_signal || state.window >= self.max {
            return;
        }
        let target = state.window + 1;
        adjust_window(&mut state, self.min, self.max, target);
    }

    pub fn on_rate_limit(&self) {
        let mut state = self
            .state
            .lock()
            .expect("adaptive semaphore mutex poisoned");
        if state.closed {
            return;
        }
        let target = (state.window / 2).max(self.min);
        adjust_window(&mut state, self.min, self.max, target);
        state.has_header_signal = false;
    }

    pub fn on_header_feedback(&self, remaining: usize, limit: usize) {
        let mut state = self
            .state
            .lock()
            .expect("adaptive semaphore mutex poisoned");
        if state.closed || limit == 0 {
            return;
        }
        state.has_header_signal = true;

        let ratio = remaining as f64 / limit as f64;
        if ratio >= HEADER_BACKOFF_THRESHOLD {
            if state.window < self.max {
                let target = state.window + 1;
                adjust_window(&mut state, self.min, self.max, target);
            }
            return;
        }

        let scale = ratio / HEADER_BACKOFF_THRESHOLD;
        let target = (self.min as f64 + scale * (self.max - self.min) as f64) as usize;
        adjust_window(&mut state, self.min, self.max, target);
    }
}

impl AdaptiveSemaphorePermit<'_> {
    pub fn release(mut self) {
        if !self.released {
            self.semaphore.release();
            self.released = true;
        }
    }
}

impl Drop for AdaptiveSemaphorePermit<'_> {
    fn drop(&mut self) {
        if !self.released {
            self.semaphore.release();
            self.released = true;
        }
    }
}

fn wake_waiter(
    waiters: &mut VecDeque<oneshot::Sender<Result<(), AdaptiveSemaphoreClosedError>>>,
) -> bool {
    while let Some(waiter) = waiters.pop_front() {
        if waiter.send(Ok(())).is_ok() {
            return true;
        }
    }
    false
}

fn adjust_window(state: &mut AdaptiveSemaphoreState, min: usize, max: usize, mut target: usize) {
    target = target.max(min).min(max);
    if target == state.window {
        return;
    }

    if target > state.window {
        let mut grow = target - state.window;
        let absorbed = grow.min(state.permits_to_absorb);
        state.permits_to_absorb -= absorbed;
        grow -= absorbed;
        for _ in 0..grow {
            if !wake_waiter(&mut state.waiters) {
                state.available += 1;
            }
        }
    } else {
        let shrink = state.window - target;
        let immediate = shrink.min(state.available);
        state.available -= immediate;
        state.permits_to_absorb += shrink - immediate;
    }

    state.window = target;
}

pub fn parse_openai_ratelimit_headers(
    headers: &HashMap<String, String>,
) -> Option<RateLimitHeaders> {
    let remaining = headers.get("x-ratelimit-remaining-requests")?;
    let limit = headers.get("x-ratelimit-limit-requests")?;

    let remaining = remaining.parse::<usize>().ok()?;
    let limit = limit.parse::<usize>().ok()?;

    (limit > 0).then_some(RateLimitHeaders { remaining, limit })
}

File: packages/llmix/python/llmix/adaptive_semaphore.py
"""AIMD Adaptive Semaphore for rate-limit-aware concurrency control.

Adjusts concurrency window dynamically:
- Additive increase on success (+1)
- Multiplicative decrease on 429 (/2)
- Preemptive backoff from rate-limit headers
"""

import asyncio
from collections import deque

DEFAULT_INITIAL = 32
DEFAULT_MIN_CONCURRENCY = 4
HEADER_BACKOFF_THRESHOLD = 0.10


class AdaptiveSemaphore:
    """Async semaphore with AIMD concurrency control and header-based early warning.

    When rate-limit headers are available (OpenAI):
      - on_header_feedback(remaining, limit): preemptive backoff when remaining < 10%.
        Above threshold -> AIMD grow as normal. Below -> scale proportionally.
    When no headers:
      - on_success(): window += 1 (additive increase)
    Always:
      - on_rate_limit(): window //= 2 (multiplicative decrease)
    """

    def __init__(
        self,
        initial: int = DEFAULT_INITIAL,
        min_concurrency: int = DEFAULT_MIN_CONCURRENCY,
    ) -> None:
        if initial < 1:
            raise ValueError(f"initial must be >= 1, got {initial}")
        if initial < min_concurrency:
            raise ValueError(
                f"initial ({initial}) must be >= min_concurrency ({min_concurrency})"
            )
        self._max = initial
        self._min = min_concurrency
        self._window = initial
        self._available = initial
        self._waiters: deque[asyncio.Future[None]] = deque()
        self._has_header_signal = False
        self._permits_to_absorb = 0

    @property
    def window(self) -> int:
        return self._window

    @property
    def max_concurrency(self) -> int:
        return self._max

    @property
    def min_concurrency(self) -> int:
        return self._min

    def rebind(self) -> None:
        """Reset loop-bound waiter state when reusing across asyncio.run() boundaries."""
        self._available = self._window
        self._waiters = deque()
        self._permits_to_absorb = 0

    async def acquire(self) -> None:
        if self._available > 0:
            self._available -= 1
            return

        fut = asyncio.get_running_loop().create_future()
        self._waiters.append(fut)
        try:
            await fut
        except Exception:
            if fut in self._waiters:
                self._waiters.remove(fut)
            elif fut.done() and not fut.cancelled():
                self.release()
            raise

    def release(self) -> None:
        if self._permits_to_absorb > 0:
            self._permits_to_absorb -= 1
            return
        if self._wake_waiter():
            return
        if self._available < self._window:
            self._available += 1

    def on_success(self) -> None:
        """AIMD additive increase -- only when no header signals are flowing."""
        if self._has_header_signal:
            return
        if self._window < self._max:
            self._adjust_window(self._window + 1)

    def on_rate_limit(self) -> None:
        """Hard 429 signal -- always halves, overrides headers."""
        new = max(self._window // 2, self._min)
        self._adjust_window(new)
        # Reset header latch so AIMD success growth resumes after the 429 storm passes
        self._has_header_signal = False

    def on_header_feedback(self, remaining: int, limit: int) -> None:
        """Header-based early warning. Backs off when remaining is scarce."""
        if limit <= 0:
            return
        self._has_header_signal = True
        ratio = remaining / limit
        if ratio >= HEADER_BACKOFF_THRESHOLD:
            # Plenty of headroom -- let AIMD grow normally
            if self._window < self._max:
                self._adjust_window(self._window + 1)
        else:
            # Scarce: scale proportionally within the danger zone
            # ratio=threshold -> full max, ratio=0 -> min_concurrency
            scale = ratio / HEADER_BACKOFF_THRESHOLD
            target = int(self._min + scale * (self._max - self._min))
            self._adjust_window(target)

    def _adjust_window(self, target: int) -> None:
        """Shared grow/shrink logic."""
        target = max(target, self._min)
        target = min(target, self._max)
        if target == self._window:
            return
        if target > self._window:
            grow = target - self._window
            absorbed = min(grow, self._permits_to_absorb)
            self._permits_to_absorb -= absorbed
            grow -= absorbed
            for _ in range(grow):
                if not self._wake_waiter():
                    self._available += 1
        else:
            shrink = self._window - target
            immediate = min(shrink, self._available)
            self._available -= immediate
            self._permits_to_absorb += shrink - immediate
        self._window = target

    def _wake_waiter(self) -> bool:
        while self._waiters:
            waiter = self._waiters.popleft()
            if waiter.cancelled():
                continue
            waiter.set_result(None)
            return True
        return False


def parse_openai_ratelimit_headers(headers: dict[str, str]) -> dict[str, int] | None:
    """Extract rate-limit info from OpenAI response headers.

    Returns {"remaining": int, "limit": int} or None if headers not present/valid.
    """
    remaining = headers.get("x-ratelimit-remaining-requests")
    limit = headers.get("x-ratelimit-limit-requests")
    if remaining is not None and limit is not None:
        try:
            rem, lim = int(remaining), int(limit)
            if lim > 0:
                return {"remaining": rem, "limit": lim}
        except (ValueError, TypeError):
            pass
    return None
